### PR TITLE
[SL-803] Scaffold parsers

### PR DESCRIPTION
**Pull request by P0lip**

### CODE REVIEW

File: src/ast-types/smdast/element.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import * as Unist from '@core/unist';
```
Comment (marbemac) [inline src/ast-types/smdast/element.ts]: Legacy `@core` import. However, not sure if we should move these ast-types over until we have a better idea how we'll use them in this package.

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+export * from './stringify';
+export * from './parse';
+
+// parser is a heavier dep, so it is not included in default exports. must import directly
+// export * from './parseWithPointers';
```
Comment (marbemac) [inline src/index.ts]: We should probably just export the parse and parseWithPointers functions as well for this library. It's the primary purpose. And take the corresponding sentence out of the readme about it not being exported from root.

File: src/parseWithPointers.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import { parse } from './parse';
+import * as Mdast from './ast-types/mdast';
+import { IPosition } from './ast-types/unist';
+import * as TUnist from './ast-types/unist';
+
+interface Pointers {
+  [key: string]: IPosition
+}
+
+const getPointers = (node: Partial<TUnist.IParent>, pointers: Pointers = {}, path: string = '/'): Pointers  => {
```
Comment (marbemac) [inline src/parseWithPointers.ts]: Should return [IParserResultPointers](https://github.com/stoplightio/types/blob/master/src/parsers.ts#L23).

File: src/parseWithPointers.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  if (node.children) {
+    for (const [i, child] of node.children.entries()) {
+      getPointers(child, pointers, `${path}children/${i}/`)
+    }
+  }
+
+  return pointers;
+};
+
+export const parseWithPointers = (value: string, opts?: ParseOpts, processor?: Processor): IParserResult<Mdast.IRoot> => {
```
Comment (marbemac) [inline src/parseWithPointers.ts]: Suggest add generic, and default to plain unist structure. Users of the library (like slate), can pass in the custom Mdast.IRoot type.

**Review by marbemac — CHANGES_REQUESTED** [review]
Looks pretty good, put in a few small suggestions. In addition:

- [x] lint errors need to be fixed, try `yarn test.prod`
- [x] formatting is incorrect, try `yarn lint.fix` (or perhaps get your IDE auto-formatting if its not?)
- [x] is it pre-mature to add the transformers and ast-types code over? should we leave it out for now and just rely on base unist types (that can be overridden w generics), and then add it later when we need to?

**Comment by P0lip** [comment]
Yeah, I had TSLint turned off for this project, that's why I had Prettier off.
I removed those transformers and indeed this is something very specific to Slate for now.

**Comment by P0lip** [comment]
@marbemac comments addressed

**Review by marbemac — APPROVED** [review]

**Pull request merged**